galgebra/ga.py

```diff
@@ -794,115 +760,132 @@ def _build_basis_blade_symbol(self, base_index):
-                for igrade, grade in enumerate(self.bases)
-                for base in grade
-            }
+        where :math:`0 < i_{1} < i_{2} < ... < i_{r}` and :math:`0 < r \le n` the
+        dimension of the vector space and :math:`0 < i_{j} \le n`. The total
+        number of all symbols of this form is :math:`2^{n}`.
+        Any multivector can be represented as a linear combination of these bases.
 
-    def _build_super_script_strings(self):
+        For the case of an orthogonal set of basis vectors the bases and blades
+        are identical, and so this attribute raises :exc:`ValueError`.
+        """
+        if self.is_ortho:
+            raise ValueError("There is no need for bases in orthogonal algebras")
+        return self.indexes._map(
+            lambda index: self._build_basis_base_symbol(index))
+
+    @_cached_property
+    def indexes_to_bases_dict(self) -> OrderedBiMap[Tuple[int, ...], Symbol]:
+        """ Bidirectional map from index tuples (:attr:`indices`) to basis bases (:attr:`bases`) """
+        return OrderedBiMap(list(zip(self.indexes.flat, self.bases.flat)))
+
+    @_cached_property
+    def bases_to_grades_dict(self) -> Dict[Symbol, int]:
+        return {
+            blade: igrade
+            for igrade, grade in enumerate(self.bases)
+            for blade in grade
+        }
+
+    @_cached_property
+    def basis_super_scripts(self):
         if self.coords is None:
             base0 = str(self.basis[0])
             if '_' in base0:
                 sub_index = base0.index('_')
-                self.basis_super_scripts = [str(base)[sub_index + 1:] for base in self.basis]
+                return [str(base)[sub_index + 1:] for base in self.basis]
             else:
-                self.basis_super_scripts = [str(i + 1) for i in self.n_range]
+                return [str(i + 1) for i in self.n_range]
         else:
-            self.basis_super_scripts = [str(coord) for coord in self.coords]
+            return [str(coord) for coord in self.coords]
 
-        self.blade_super_scripts = self.indexes._map(lambda base_index: ''.join(
+    @_cached_property
+    def blade_super_scripts(self):
+        return self.indexes._map(lambda base_index: ''.join(
             self.basis_super_scripts[i] for i in base_index
         ))
 
-    def _build_mv_wrappers(self):
-        # create the Mv wrappers
-        self.mv_blades = self.blades._map(lambda blade: mv.Mv(blade, ga=self))
+    @_cached_property
+    def mv_blades(self) -> GradedTuple[_mv.Mv]:
+        """ :class:`mv.Mv` instances corresponding to :attr:`blades`. """
+        return self.blades._map(lambda blade: mv.Mv(blade, ga=self))
 
-        self.mv_basis = tuple(
-            mv.Mv(obj, ga=self)
-            for obj in self.basis
-        )
+    @_cached_property
+    def mv_basis(self) -> Tuple[_mv.Mv]:
+        """ :class:`mv.Mv` instances corresponding to :attr:`basis`. """
+        return tuple(mv.Mv(obj, ga=self) for obj in self.basis)
```

Comment:
This is a fairly simple example of the type of code this results in. A nice feature is the reader no longer needs to think about thing like "is this variable initialized yet", as the properties result in an appropriate computation order automatically. Of course, if you end up with a cycle then you'll have a bad time.